File: chapters/01-fire/rust/src/main.rs
use raylib_ffi::{*, colors::WHITE};
use std::ffi::c_void;
use tinyrand::*;

fn main() {
    init_gui();

    let mut screen_buffer_data = [ Color{r:0,g:0,b:0,a:0xFF} ; 400*300];
    let screen_buffer = Image {
        data: screen_buffer_data.as_mut_ptr() as *mut c_void,
        width: 400,
        height: 300,
        format: enums::PixelFormat::R8g8b8a8 as i32,
        mipmaps: 1,
    };
    let screen_buffer_texture 
        = load_texture_from_image(screen_buffer);
    let palette = generate_palette();
    let mut fire_buffer = [ 0u8; 400*300 ];
    let mut rng = StdRand::default();

    while ! window_should_close() {
        begin_drawing();

        draw_next_frame(&mut screen_buffer_data, &mut fire_buffer, &palette, &mut rng);

        // Pasar a la GPU
        update_texture(
            screen_buffer_texture, 
            &screen_buffer_data
        );
        draw_texture(screen_buffer_texture, 0, 0, WHITE);
        end_drawing()
    }
    println!("Hello, world!");

    close_window();
}

fn generate_palette() -> [Color; 256] {
    let mut pal = [ Color{r:0,g:0,b:0,a:0xFF} ; 256];

    for i in 0..=84 {
        pal[i].r = (i * (0xFF / 85)) as u8;
    }
    for i in 85..=169 {
        pal[i].r = 255;
        pal[i].g = ((i-85) * (0xFF / 85)) as u8;
    }
    for i in 170..=255 {
        pal[i].r = 255;
        pal[i].g = 255;
        pal[i].b = ((i-170) * (0xFF / 85)) as u8;
    }

    pal
}

fn draw_the_palette(screen: &mut [Color], pal: &[Color]) {
    for i in 0..pal.len() {
        let init = i*400 + 50;
        let pixeles = &mut screen[ init..(init+4) ];
        pixeles[0] = pal[i];
        pixeles[1] = pal[i];
        pixeles[2] = pal[i];
        pixeles[3] = pal[i];
    }
}

fn fill_bottom_with_random_ashes(fire_buf: &mut[u8], rng: &mut Wyrand) {
    let end   = fire_buf.len();
    let start = end - 400;
    for i in start..end {
        fire_buf[i] = rng.next_range(0..256u16) as u8;
    }
}

fn convert_fire_buffer_to_screen(fire_buf: &[u8], pal: &[Color], screen: &mut [Color]) {
    for i in 0..fire_buf.len() {
        let heat = fire_buf[i] as usize;
        screen[i] = pal[ heat ];
    }
}

fn calculate_next_fire_frame(fire_buf: &mut[u8]) {
    let mut old_fire_buf = [0u8; 400*300];
    old_fire_buf.clone_from_slice(fire_buf);

    for y in 0..299 {
        for x in 1..399 {
            let i = (y*400 + x) as usize;

            fire_buf[i] = ((
                    10 * old_fire_buf[i - 1] as u64
                +   20 * old_fire_buf[i + 0] as u64
                +   10 * old_fire_buf[i + 1] as u64
                +  160 * old_fire_buf[i - 1 + 400] as u64
                +  320 * old_fire_buf[i + 0 + 400] as u64
                +  160 * old_fire_buf[i + 1 + 400] as u64
            ) / 680) as u8;
        }
    }
}

fn draw_next_frame(screen: &mut [Color], fire_buf: &mut[u8], pal: &[Color], rng: &mut Wyrand) {
    draw_the_palette(screen, pal);
    fill_bottom_with_random_ashes(fire_buf, rng);
    calculate_next_fire_frame(fire_buf);
    convert_fire_buffer_to_screen(fire_buf, pal, screen);
}

fn update_texture(tex: Texture2D, data: &[Color]) {
    unsafe{ 
        UpdateTexture(tex, data.as_ptr() as *const c_void);
    }
}

fn draw_texture(tex: Texture2D, x: i32, y: i32, tint: Color) {
    unsafe {
        DrawTexture(tex, x, y, tint);
    }
}

fn load_texture_from_image(img: Image) -> Texture2D {
    unsafe {
        LoadTextureFromImage(img)
    }
}

fn init_gui() {
    unsafe {
        InitWindow(400, 300, "Fire".as_ptr() as *const i8);
    }
}

fn window_should_close() -> bool {
    unsafe { WindowShouldClose() }    
}

fn begin_drawing() {
    unsafe {
        BeginDrawing();
    }    
}

fn end_drawing() {
    unsafe {
        EndDrawing();
    }    
}

fn close_window() {
    unsafe {
        CloseWindow();   
    }
}

fn draw_pixel(x: i32, y: i32, color: Color) {
    unsafe {
        DrawPixel(x, y, color)
    }
}

File: chapters/01-fire/c/main.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <raylib.h>

void calculate_next_fire_frame(uint8_t fire_buffer[400*300]) {

    uint8_t old_fire_buffer[400*300] = {};
    memcpy(old_fire_buffer, fire_buffer, (size_t)400*300);

    for(int y=1; y < (300-1); y++) {
        int p = y*400;
        for(int x=1; x < (400-1); x++) {
            int i = p+x;
            fire_buffer[i] = (uint8_t)((
               10 * old_fire_buffer[i - 1]
            +  20 * old_fire_buffer[i + 0]
            +  10 * old_fire_buffer[i + 1]
            + 160 * old_fire_buffer[i - 1 + 400]
            + 320 * old_fire_buffer[i + 0 + 400]
            + 160 * old_fire_buffer[i + 1 + 400]
            ) / 680);
        }
    }
    
}

void convert_fire_buffer_to_screen(Image screen_buffer, Color palette[256], uint8_t fire_buffer[400*300]) {
    Color *pixel = screen_buffer.data;
    for(int i = 0; i < 400 * 300; i++) {
        pixel[i] = palette[fire_buffer[i]];
    }
}

void fill_bottom_with_random_ashes(uint8_t fire_buffer[400*300]) {
    for(int i = 0; i < 400; i++) {
        fire_buffer[400*299 + i] = (uint8_t)rand() % 256;
    }
}

void draw_next_frame(Image screen_buffer, Color palette[256], uint8_t fire_buffer[400*300]) {
    fill_bottom_with_random_ashes(fire_buffer);
    calculate_next_fire_frame(fire_buffer);
    convert_fire_buffer_to_screen(screen_buffer, palette, fire_buffer);
}

void generate_palette(Color palette[256]) {
    for(int i = 0; i <= 84; i++) {
        uint8_t red = i * (0xFF / 84);
        palette[i] = (Color){.r=red,.a=0xFF};
    }

    for(int i = 85; i <= 169; i++) {
        uint8_t green = (i - 85) * (0xFF / 84);
        palette[i] = (Color){.r=0xFF,.g=green,.a=0xFF};
    }

    for(int i = 170; i <= 256; i++) {
        uint8_t blue = (i - 170) * (0xFF / 85);
        palette[i] = (Color){.r=0xFF,.g=0xFF,.b=blue,.a=0xFF};
    }
}

int main() {
    InitWindow(400, 300, "HELLO WORLD!");

    Color palette[256] = {};
    generate_palette(palette);

    // El error que se vió en directo fue no inicializar esta variable :s
    // La línea original era:
    // unsigned char fire_buffer[400*300];
    uint8_t fire_buffer[400*300] = {};

    Color screen_buffer_data[400*300] = {};

    Image screen_buffer = {
        .data = screen_buffer_data,
        .width = 400,
        .height = 300,
        .format = PIXELFORMAT_UNCOMPRESSED_R8G8B8A8,
        .mipmaps = 1
    };

    Texture2D screen_buffer_texture = LoadTextureFromImage(screen_buffer);

    while(!WindowShouldClose()) {
        BeginDrawing();

        draw_next_frame(screen_buffer, palette, fire_buffer);

        // Pasar a la GPU
        UpdateTexture(screen_buffer_texture, screen_buffer.data);
        DrawTexture(screen_buffer_texture, 0, 0, WHITE);
        //DrawPixel(200, 200, (Color){.r=255,.g=0,.b=0,.a=255});

        EndDrawing();
    }

    CloseWindow();
    return 0;
}